posts/2023-06-07-kotlin-ktor-keycloak.adoc

```diff
@@ -0,0 +1,357 @@
+
+In order to query the `luna-vacation-api` we need a http client. We are going to use the https://ktor.io/docs/create-client.html[ktor-client] library by adding the necessary dependencies to `build.gradle.kts`:
+
+[source,kotlin]
+----
+implementation("io.ktor:ktor-client-core:2.3.1")
+implementation("io.ktor:ktor-client-cio:2.3.1")
+implementation("io.ktor:ktor-client-serialization:2.3.1")
+implementation("io.ktor:ktor-client-content-negotiation:2.3.1")
+implementation("io.ktor:ktor-serialization-kotlinx-json:2.3.1")
+----
+
+Besides creating the http client we also need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization,
+where we will ignore the json fields that we are not interested in.
+
+[source,kotlin]
+----
+    val client = HttpClient(CIO) {
+        install(ContentNegotiation) {
+            json(
+                Json {
+                    ignoreUnknownKeys = true
+                },
+            )
+        }
+    }
+----
+
+Now that the http client is properly setup we will start by retrieving a JWT token from Keycloak:
+
+[source,kotlin]
+----
+@Serializable
+data class BotToken(val access_token: String)
+
+val token = client.submitForm(
+        url = "http://localhost:8080/realms/lunarealm/protocol/openid-connect/token",
+        formParameters = parameters {
+            append("grant_type", "client_credentials")
+            append("client_id", "luna-vacation-bot")
+            append("client_secret", "<replace by client_secret>")
+        },
+    ).body<BotToken>().access_token
+----
+
+and then we can finally call the `luna-vacation-api`:
+
+[source,kotlin]
+----
+val response = client.get("http://localhost:5050/imionvacation") {
+        bearerAuth(token)
+    }.bodyAsText()
+----
+
+And what do we do with the `response` now? We send it to https://slack.com/[Slack] using a https://slack.com/apps[Slack App]. So let's do that next.
+
+
+=== Create a Slack app
+You can create your own workspace at https://slack.com/[slack.com]. Once you have it you can create a new https://api.slack.com/apps[App]. Choose `Create New App` and then `From scratch`:
+
+.Create a new Slack App
+image::slack_create_app_1.png[Diagram, 400]
+
+I named the app `ImIOnVacation` and added it to my personal workspace:
+
+.Configure a Slack App worskpace
+image::slack_create_app_2.png[Diagram, 450]
+
+To finalize the App creation you need to set some permissions:
+
+.Setup Slack App permissions
+image::slack_app_permissions.png[Diagram, 400]
+
+In `Basic Information` you can personalize your app appearance, but for now let's jump to `OAuth & Permission` and, from there, scroll down to the `Scopes` section. In `User Token Scopes` add the scopes `chat:write`:
+
+.Setup User Token Scopes
+image::slack_app_scopes.png[Diagram, 500]
+
+Then scroll up and click `Install to Workspace`. That will take you to a permissions screen: click `Allow`. You can see that a `User OAuth Token` was created as well.
```

Comment:
```suggestion
Then scroll up and click `Install to Workspace`. That will take us to a permissions screen: click `Allow`. We can see that a `User OAuth Token` was created as well.
```

`you` or `we`? For the sake of consistency, I'd say use first-person pronouns throughout the blog post.